packages/ai/core/generate-text/run-tools-transformation.ts

```diff
@@ -133,6 +135,18 @@ export function runToolsTransformation<TOOLS extends ToolSet>({
         SingleRequestTextStreamPart<TOOLS>
       >,
     ) {
+      // Initialize dataStream that writes to the forward stream controller
+      const dataStream: BasicDataStreamWriter = {
+        writeData(data: unknown) {
+          controller.enqueue(formatStreamPart('data', [data]));
+        },
+        writeMessageAnnotation(annotation: unknown) {
+          controller.enqueue(
+            formatStreamPart('message_annotations', [annotation]),
+          );
+        },
+      };
+
```

Comment:
at this stage we do not have a data stream so this goes against the architecture of streamText. agree with the general idea though - but the integration will not be that simple (thought about it quite a bit)

Reply:
Ah; yeah I was looking at the ai chatbot over the weekend and how it handles custom data. 

Realized being able to write to a custom data stream would be pretty useful for execute and onToolCall, especially for longer tool calls with an api that supports streams.

Tried to find a good DX but yeah couldn’t exactly find best way to integrate data stream into stream text.